File: ios/new/ExperimentalAssetLoader.swift
@_spi(RiveExperimental) import RiveRuntime
import NitroModules

enum AssetType {
  case image
  case font
  case audio

  init?(fromName name: String) {
    let lowercased = name.lowercased()
    if lowercased.contains("image") || lowercased.hasSuffix(".png") || lowercased.hasSuffix(".jpg") || lowercased.hasSuffix(".jpeg") || lowercased.hasSuffix(".webp") {
      self = .image
    } else if lowercased.contains("font") || lowercased.hasSuffix(".ttf") || lowercased.hasSuffix(".otf") {
      self = .font
    } else if lowercased.contains("audio") || lowercased.hasSuffix(".wav") || lowercased.hasSuffix(".mp3") || lowercased.hasSuffix(".flac") || lowercased.hasSuffix(".ogg") {
      self = .audio
    } else {
      return nil
    }
  }
}

@MainActor
final class ExperimentalAssetLoader {

  static func registerAssets(
    _ referencedAssets: ReferencedAssetsType?,
    on worker: Worker
  ) async {
    guard let assets = referencedAssets?.data else { return }

    await withTaskGroup(of: Void.self) { group in
      for (name, asset) in assets {
        group.addTask { @MainActor in
          await self.loadAndRegisterAsset(name: name, asset: asset, worker: worker)
        }
      }
    }
  }

  private static func loadAndRegisterAsset(
    name: String,
    asset: ResolvedReferencedAsset,
    worker: Worker
  ) async {
    do {
      let data = try await loadAssetData(asset)
      guard !data.isEmpty else { return }

      let assetType = AssetType(fromName: name) ?? inferAssetType(from: asset, data: data)
      guard let assetType = assetType else {
        RCTLogWarn("Could not determine asset type for: \(name)")
        return
      }

      try await registerAsset(data: data, name: name, type: assetType, worker: worker)
    } catch {
      RCTLogError("Failed to load asset '\(name)': \(error)")
    }
  }

  private static func loadAssetData(_ asset: ResolvedReferencedAsset) async throws -> Data {
    guard let dataSource = try DataSourceResolver.resolve(from: asset) else {
      return Data()
    }
    return try await dataSource.createLoader().load(from: dataSource)
  }

  private static func inferAssetType(from asset: ResolvedReferencedAsset, data: Data) -> AssetType? {
    if let sourceUrl = asset.sourceUrl {
      if let type = AssetType(fromName: sourceUrl) {
        return type
      }
    }
    if let sourceAsset = asset.sourceAsset {
      if let type = AssetType(fromName: sourceAsset) {
        return type
      }
    }
    if let sourceAssetId = asset.sourceAssetId {
      if let type = AssetType(fromName: sourceAssetId) {
        return type
      }
    }

    // Try to infer from data magic bytes
    if data.count >= 4 {
      let bytes = [UInt8](data.prefix(4))
      // PNG: 89 50 4E 47
      if bytes[0] == 0x89 && bytes[1] == 0x50 && bytes[2] == 0x4E && bytes[3] == 0x47 {
        return .image
      }
      // JPEG: FF D8 FF
      if bytes[0] == 0xFF && bytes[1] == 0xD8 && bytes[2] == 0xFF {
        return .image
      }
      // WebP: RIFF....WEBP
      if bytes[0] == 0x52 && bytes[1] == 0x49 && bytes[2] == 0x46 && bytes[3] == 0x46 {
        return .image
      }
    }

    return nil
  }

  private static func registerAsset(
    data: Data,
    name: String,
    type: AssetType,
    worker: Worker
  ) async throws {
    RCTLogInfo("ExperimentalAssetLoader: Registering \(type) asset '\(name)' (\(data.count) bytes)")
    switch type {
    case .image:
      worker.removeGlobalImageAsset(name: name)
      let image = try await worker.decodeImage(from: data)
      worker.addGlobalImageAsset(image, name: name)
      RCTLogInfo("ExperimentalAssetLoader: Image '\(name)' registered successfully")

    case .font:
      worker.removeGlobalFontAsset(name)
      let font = try await worker.decodeFont(from: data)
      worker.addGlobalFontAsset(font, name: name)
      RCTLogInfo("ExperimentalAssetLoader: Font '\(name)' registered successfully")

    case .audio:
      worker.removeGlobalAudioAsset(name: name)
      let audio = try await worker.decodeAudio(from: data)
      worker.addGlobalAudioAsset(audio, name: name)
      RCTLogInfo("ExperimentalAssetLoader: Audio '\(name)' registered successfully")
    }
  }
}

File: android/src/experimental/java/com/margelo/nitro/rive/ExperimentalAssetLoader.kt
package com.margelo.nitro.rive

import android.util.Log
import app.rive.AudioAsset
import app.rive.FontAsset
import app.rive.ImageAsset
import app.rive.core.CommandQueue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

object ExperimentalAssetLoader {
  private const val TAG = "ExperimentalAssetLoader"

  fun registerAssets(
    referencedAssets: ReferencedAssetsType?,
    riveWorker: CommandQueue
  ) {
    val assetsData = referencedAssets?.data ?: return
    val scope = CoroutineScope(Dispatchers.IO)

    for ((name, assetData) in assetsData) {
      val source = DataSourceResolver.resolve(assetData) ?: continue
      scope.launch {
        try {
          val loader = source.createLoader()
          val data = loader.load(source)
          val type = inferAssetType(name, data)
          registerAsset(data, name, type, riveWorker)
        } catch (e: Exception) {
          Log.e(TAG, "Failed to load asset '$name'", e)
        }
      }
    }
  }

  fun updateAssets(
    referencedAssets: ReferencedAssetsType,
    riveWorker: CommandQueue
  ) {
    val assetsData = referencedAssets.data ?: return
    val scope = CoroutineScope(Dispatchers.IO)

    for ((name, assetData) in assetsData) {
      val source = DataSourceResolver.resolve(assetData) ?: continue
      scope.launch {
        try {
          val loader = source.createLoader()
          val data = loader.load(source)
          val type = inferAssetType(name, data)
          registerAsset(data, name, type, riveWorker)
        } catch (e: Exception) {
          Log.e(TAG, "Failed to update asset '$name'", e)
        }
      }
    }
  }

  private suspend fun registerAsset(
    data: ByteArray,
    name: String,
    type: AssetType,
    riveWorker: CommandQueue
  ) {
    Log.i(TAG, "Registering $type asset '$name' (${data.size} bytes)")
    when (type) {
      AssetType.IMAGE -> {
        riveWorker.unregisterImage(name)
        val result = ImageAsset.fromBytes(riveWorker, data)
        if (result is app.rive.Result.Success) {
          result.value.register(name)
          Log.i(TAG, "Image '$name' registered")
        }
      }
      AssetType.FONT -> {
        riveWorker.unregisterFont(name)
        val result = FontAsset.fromBytes(riveWorker, data)
        if (result is app.rive.Result.Success) {
          result.value.register(name)
          Log.i(TAG, "Font '$name' registered")
        }
      }
      AssetType.AUDIO -> {
        riveWorker.unregisterAudio(name)
        val result = AudioAsset.fromBytes(riveWorker, data)
        if (result is app.rive.Result.Success) {
          result.value.register(name)
          Log.i(TAG, "Audio '$name' registered")
        }
      }
    }
  }

  private fun inferAssetType(name: String, data: ByteArray): AssetType {
    val ext = name.substringAfterLast('.', "").lowercase()
    return when (ext) {
      "png", "jpg", "jpeg", "webp", "gif", "bmp", "svg" -> AssetType.IMAGE
      "ttf", "otf", "woff", "woff2" -> AssetType.FONT
      "wav", "mp3", "ogg", "flac", "aac", "m4a" -> AssetType.AUDIO
      else -> inferFromMagicBytes(data)
    }
  }

  private fun inferFromMagicBytes(data: ByteArray): AssetType {
    if (data.size < 4) return AssetType.IMAGE

    // PNG: 89 50 4E 47
    if (data[0] == 0x89.toByte() && data[1] == 0x50.toByte() &&
      data[2] == 0x4E.toByte() && data[3] == 0x47.toByte()) return AssetType.IMAGE
    // JPEG: FF D8 FF
    if (data[0] == 0xFF.toByte() && data[1] == 0xD8.toByte() &&
      data[2] == 0xFF.toByte()) return AssetType.IMAGE
    // RIFF container: WebP (RIFF....WEBP) or WAV (RIFF....WAVE)
    if (data[0] == 0x52.toByte() && data[1] == 0x49.toByte() &&
      data[2] == 0x46.toByte() && data[3] == 0x46.toByte()) {
      if (data.size >= 12 &&
        data[8] == 0x57.toByte() && data[9] == 0x41.toByte() &&
        data[10] == 0x56.toByte() && data[11] == 0x45.toByte()) return AssetType.AUDIO // "WAVE"
      return AssetType.IMAGE // assume WebP for other RIFF
    }
    // ID3 (MP3): 49 44 33
    if (data[0] == 0x49.toByte() && data[1] == 0x44.toByte() &&
      data[2] == 0x33.toByte()) return AssetType.AUDIO
    // TrueType: 00 01 00 00
    if (data[0] == 0x00.toByte() && data[1] == 0x01.toByte() &&
      data[2] == 0x00.toByte() && data[3] == 0x00.toByte()) return AssetType.FONT
    // OpenType: 4F 54 54 4F ("OTTO")
    if (data[0] == 0x4F.toByte() && data[1] == 0x54.toByte() &&
      data[2] == 0x54.toByte() && data[3] == 0x4F.toByte()) return AssetType.FONT

    return AssetType.IMAGE
  }

  enum class AssetType { IMAGE, FONT, AUDIO }
}
